### PR TITLE
using Violation object instead of strings

### LINKED ISSUE #827: Rules should not return `?string`, but a `Violation` object

### PR DESCRIPTION
**Pull request by AlessandroMinoccheri**
This should fix #827 

I added an interface `ViolationInterface` with the NullObject pattern to avoid multiple return types.
I know that the PR is really big but I couldn't split it into many parts.

### CODE REVIEW

**Comment by OskarStark** [comment]
I will have a look next week after my vacation. Many thanks 😃

**Review by OskarStark — APPROVED** [review]
Looks good from my side, can you please double check @alamirault

**Comment by OskarStark** [comment]
Please rebase, thanks

**Comment by AlessandroMinoccheri** [comment]
Rebased @OskarStark

**Comment by OskarStark** [comment]
Can you please check the CI?

**Comment by AlessandroMinoccheri** [comment]
@OskarStark There is a step `Documentation/Generate ` that is not passing.
What can I do?

**Comment by OskarStark** [comment]
> @OskarStark There is a step Documentation/Generate  that is not passing.
> What can I do?

I am not sure, you can try to rebase once again, I tried to fix it, not sure its working on a PR from a remote repository.

**Comment by AlessandroMinoccheri** [comment]
The pull request is up to date with develop branch.
I think that the step can't be run from other repositories (forks)

**Comment by OskarStark** [comment]
Thank you 🙏

**Comment by OskarStark** [comment]
Hmm, after this PR, the whitelisted violations appear 🤔 

See https://github.com/symfony/symfony-docs/actions/runs/4045272667/jobs/6956522269

cc @alamirault

**Comment by OskarStark** [comment]
I pinned symfony-docs to 1.35.1
* https://github.com/symfony/symfony-docs/pull/17830

**Comment by alamirault** [comment]
> Hmm, after this PR, the whitelisted violations appear thinking
> 
> See https://github.com/symfony/symfony-docs/actions/runs/4045272667/jobs/6956522269
> 
> cc @alamirault

I can't found origin, I will check it this weekend

File: src/Analyzer/RstAnalyzer.php [hunk 1]
```diff
@@ -107,15 +97,10 @@ public function analyze(\SplFileInfo $file, array $rules): array
...
                 }
 
-                $violationMessage = $rule->check($lines, $no);
+                $violation = $rule->check($lines, $no, (string) $file->getRealPath());
 
-                if (null !== $violationMessage) {
-                    $violations[] = Violation::from(
-                        $violationMessage,
-                        (string) $file->getRealPath(),
-                        $no + 1,
```
Comment (OskarStark) [inline src/Analyzer/RstAnalyzer.php:116]: Before we had here `$no + 1`, but afterwards, all LineContentRules are using:
```php
           return Violation::from(
                 'a message'
                 $filename,
                 1,
                 ''
             );
```

And Line number 1, or am I missing sth @AlessandroMinoccheri 🤔

Comment (OskarStark) [reply src/Analyzer/RstAnalyzer.php:116]: Also the raw line is missing 🤔

Comment (OskarStark) [reply src/Analyzer/RstAnalyzer.php:116]: Spotted while working on https://github.com/OskarStark/doctor-rst/pull/1276

Comment (OskarStark) [reply src/Analyzer/RstAnalyzer.php:116]: It would be awesome if you can restore the number and the raw line for the violations in a new PR 🙏

**Pull request merged**